core/endorser/endorser.go

```diff
@@ -472,86 +473,124 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
-				collections[cprws.CollectionName] = false
-			}
-			chaincodes[prws.Namespace] = collections
+	// There might be public states that are read and not written.  Need to add these to the policyRequired structure.
+	for _, nsrws := range simResult.PubSimulationResults.GetNsRwset() {
+		if e.Support.IsSysCC(nsrws.Namespace) {
+			// skip system chaincodes
+			continue
+		}
+		if _, ok := policies.policyRequired[nsrws.Namespace]; !ok {
+			// There's a public RWset for this namespace, but no public or private writes, so chaincode policy is required.
+			policies.add(nsrws.Namespace, "", true)
 		}
 	}
 
-	// 3. Add/merge collections from the PrivateReads struct
-	for ns, entry := range simResult.PrivateReads {
-		if e.Support.IsSysCC(ns) {
+	// There might be private collection states that are read and not written.  Need to add these to the policyRequired structure.
+	for chaincode, entry := range simResult.PrivateReads {
+		if e.Support.IsSysCC(chaincode) {
 			// skip system chaincodes
 			continue
 		}
-		if cc := chaincodes[ns]; cc == nil {
-			// chaincode either not already in the map, or nil entry - add it
-			chaincodes[ns] = map[string]bool{}
-		}
-		for name := range entry {
-			// even if this already exists from the write-set, it will override the value with true to indicate it was read
-			chaincodes[ns][name] = true
+		for collection := range entry {
+			if !policies.requireCollectionPolicy(chaincode, collection) {
+				policies.add(chaincode, "", true)
+			}
 		}
 	}
 
-	// now build the chaincode interest
 	ccInterest := &pb.ChaincodeInterest{}
-	for chaincode, collections := range chaincodes {
-		if collections == nil {
-			ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
+	for chaincode, collections := range policies.policyRequired {
+		for collection := range collections {
+			ccCall := &pb.ChaincodeCall{
 				Name: chaincode,
-			})
-		} else {
-			for collectionName, isRead := range collections {
+			}
+			if collection == "" { // the empty collection name here represents the public RWset
+				keyPolicies := policies.sbePolicies[chaincode]
+				if len(keyPolicies) > 0 {
+					// For simplicity, we'll always add the SBE policies to the public ChaincodeCall, and set the disregard flag if the chaincode policy is not required.
+					ccCall.KeyPolicies = keyPolicies
+					if !policies.requireChaincodePolicy(chaincode) {
+						ccCall.DisregardNamespacePolicy = true
+					}
+				}
+			} else {
 				// Since each collection in a chaincode could have different values of the NoPrivateReads flag, create a new Chaincode entry for each.
-				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
-					Name:            chaincode,
-					CollectionNames: []string{collectionName},
-					NoPrivateReads:  !isRead,
-				})
+				ccCall.CollectionNames = []string{collection}
+				ccCall.NoPrivateReads = !simResult.PrivateReads.Exists(chaincode, collection)
```

Comment:
PrivateReads may have additional entries, so you would need to loop through at the end and add the entries that are not present in the policies map.

Reply:
I'm struggling to understand this one.  If there are entries here that are not in the policies map, then it's for private reads with no writes.  But they got added to the policy map due to the existence of the hashed reads in the public RWset (as per your previous comment).  And my observation is that there require the chaincode policy.  There's an integration test for this: https://github.com/andrew-coleman/fabric/blob/DisregardNamespacePolicy/integration/gateway/gateway_discovery_test.go#L546
I'll push my latest commit for now and we can revisit this.

Reply:
Fixed in the latest patch.